File: Graphs/Detect-Cycle-Undirected-Graph-DFS/DetectCycleDFS.py
def dfs(node, parent, visited, graph):
    visited[node] = True
    for neighbor in graph[node]:
        if not visited[neighbor]:
            if dfs(neighbor, node, visited, graph):
                return True
        elif neighbor != parent:
            return True
    return False

def has_cycle(V, graph):
    visited = [False] * V
    for i in range(V):
        if not visited[i]:
            if dfs(i, -1, visited, graph):
                return True
    return False

if __name__ == "__main__":
    V = 4
    graph = [[] for _ in range(V)]

    # Example graph: 0-1, 1-2, 2-0
    graph[0].append(1)
    graph[1].append(0)

    graph[1].append(2)
    graph[2].append(1)

    graph[2].append(0)
    graph[0].append(2)

    print("Cycle exists" if has_cycle(V, graph) else "No cycle")

File: Graphs/Detect-Cycle-Undirected-Graph-DFS/DetectCycleDFS.java
import java.util.*;

public class DetectCycleDFS {
    public static boolean dfs(int node, int parent, boolean[] visited, List<List<Integer>> graph) {
        visited[node] = true;

        for (int neighbor : graph.get(node)) {
            if (!visited[neighbor]) {
                if (dfs(neighbor, node, visited, graph))
                    return true;
            } else if (neighbor != parent) {
                return true; // Found cycle
            }
        }
        return false;
    }

    public static boolean hasCycle(int V, List<List<Integer>> graph) {
        boolean[] visited = new boolean[V];
        for (int i = 0; i < V; i++) {
            if (!visited[i]) {
                if (dfs(i, -1, visited, graph))
                    return true;
            }
        }
        return false;
    }

    public static void main(String[] args) {
        int V = 4;
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < V; i++) {
            graph.add(new ArrayList<>());
        }

        // Example graph: 0-1, 1-2, 2-0
        graph.get(0).add(1);
        graph.get(1).add(0);

        graph.get(1).add(2);
        graph.get(2).add(1);

        graph.get(2).add(0);
        graph.get(0).add(2);

        System.out.println(hasCycle(V, graph) ? "Cycle exists" : "No cycle");
    }
}
